command/agent/srv.go

```diff
@@ -0,0 +1,114 @@
+package agent
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"time"
+)
+
+const (
+	srvPollInterval = 60 * time.Second
+	srvFindTimeout  = srvPollInterval / 2
+)
+
+// AgentSRV periodically polls an SRV record
+// And attempts to join the hosts supplied
+type AgentSRV struct {
+	agent      *Agent
+	srvrecords []string
+	logger     *log.Logger
+	replay     bool
+}
+
+// NewAgentSRV is used to create a new AgentSRV
+func NewAgentSRV(agent *Agent, logOutput io.Writer, replay bool, srvrecords []string) (*AgentSRV, error) {
+
+	// Initialize the AgentSRV
+	m := &AgentSRV{
+		agent:      agent,
+		srvrecords: srvrecords,
+		logger:     log.New(logOutput, "", log.LstdFlags),
+		replay:     replay,
+	}
+
+	// Start the poller in the background
+	go m.run()
+	return m, nil
+}
+
+// run is a long running goroutine that scans for new hosts periodically
+func (m *AgentSRV) run() {
+
+	for {
+		// Set up a channel and select so we can timeout if findSRV takes too long
+		// set channel to 1 so that goroutines can't pile-up
+		c := make(chan []string, 1)
+		go func() { c <- m.findSRV() }()
+		select {
+		case records := <-c:
+			// Attempt the join only if there are new records
+			if len(records) > 0 {
+				n, err := m.agent.Join(records, m.replay)
+
+				if err != nil {
+					m.logger.Printf("[ERR] agent.srv: Failed to join: %v", err)
+				}
+				if n > 0 {
+					m.logger.Printf("[INFO] agent.srv: Joined %d hosts", n)
+				}
+			}
+		// Report the timeout
+		case <-time.After(srvFindTimeout):
+			m.logger.Printf("[ERR] agent.srv: findSRV timed out")
+		}
+		// Sleep until it's time to poll again
+		time.Sleep(srvPollInterval)
+	}
+}
+
+// findSRV looks up the SRV records and returns a slice of all SRV records
+// that are not currently cluster members
+func (m *AgentSRV) findSRV() []string {
+	var hosts []string
+
+	// map the members so that we only do a single O(n) search through members
+	known_members := make(map[string]bool)
```

Comment:
You are right I was looking at the private member. Making a local map seems fine. 

> On Sep 10, 2015, at 5:41 PM, Dale Hamel <notifications@github.com> wrote:
> 
> In command/agent/srv.go:
> 
> > +		// Report the timeout
> > +		case <-time.After(srvFindTimeout):
> > +			m.logger.Printf("[ERR] agent.srv: findSRV timed out")
> > +		}
> > +		// Sleep until it's time to poll again
> > +		time.Sleep(srvPollInterval)
> > +	}
> > +}
> > +
> > +// findSRV looks up the SRV records and returns a slice of all SRV records
> > +// that are not currently cluster members
> > +func (m *AgentSRV) findSRV() []string {
> > +	var hosts []string
> > +
> > +	// map the members so that we only do a single O(n) search through members
> > +	known_members := make(map[string]bool)
> I see what you mean: members map[string]*memberState, though unfortunately members is not exported. It feels a bit silly to serialize it to a slice only to put it back to a map.
> 
> I suppose I could add a RawMembers function to avoid this, but I'm not familiar with a way to do something like a 'const' return in go (to prevent the caller from modifying the value). It scares me to do something like this, because members is clearly intended to be private, and while I only need read only access I don't know of a way to ensure this.
> 
> —
> Reply to this email directly or view it on GitHub.
>